File: java-workshop/part-3-mocks-clean/src/main/java/service/OrderCancellationService.java
package service;

import domain.CancellationResult;
import domain.Order;
import domain.OrderStatus;
import domain.StockReservation;
import repository.OrderRepository;

import java.util.List;

public class OrderCancellationService {
    private final OrderRepository orderRepository;
    private final PaymentService paymentService;
    private final StockService stockService;
    private final ShippingService shippingService;

    public OrderCancellationService(
        OrderRepository orderRepository,
        PaymentService paymentService,
        StockService stockService,
        ShippingService shippingService
    ) {
        this.orderRepository = orderRepository;
        this.paymentService = paymentService;
        this.stockService = stockService;
        this.shippingService = shippingService;
    }

    public CancellationResult cancelOrder(String orderId) {
        Order order = orderRepository.findById(orderId);
        if (order == null) {
            return new CancellationResult.Failure("Order not found");
        }

        if (order.status() != OrderStatus.CONFIRMED) {
            return new CancellationResult.Failure("Order cannot be cancelled");
        }

        if (order.transactionId() != null) {
            // THE TRAP: treats false as ERROR, but PaymentService means "nothing to refund"
            if (!paymentService.refundPayment(order.transactionId())) {
                return new CancellationResult.Failure("Refund failed");
            }
        }

        if (order.trackingNumber() != null && !shippingService.cancelShipment(order.trackingNumber())) {
            return new CancellationResult.Failure("Shipment cancellation failed");
        }

        List<StockReservation> reservations = order.items().stream()
            .map(item -> new StockReservation(item.productId(), item.quantity(), true))
            .toList();
        stockService.releaseStock(reservations);
        orderRepository.updateStatus(orderId, OrderStatus.CANCELLED);
        return new CancellationResult.Success();
    }
}

File: kotlin-workshop/part-3-mocks-clean/src/main/kotlin/service/OrderCancellationService.kt
package service

import domain.CancellationResult
import domain.OrderStatus
import domain.StockReservation
import repository.OrderRepository

/**
 * Service for cancelling orders.
 * Handles refund, stock release, and shipment cancellation.
 */
class OrderCancellationService(
    private val orderRepository: OrderRepository,
    private val paymentService: PaymentService,
    private val stockService: StockService,
    private val shippingService: ShippingService
) {

    fun cancelOrder(orderId: String): CancellationResult {
        // 1. Find the order
        val order = orderRepository.findById(orderId)
            ?: return CancellationResult.Failure("Order not found")

        // 2. Check if cancellation is allowed
        if (order.status == OrderStatus.SHIPPED) {
            return CancellationResult.Failure("Cannot cancel shipped order")
        }
        if (order.status == OrderStatus.CANCELLED) {
            return CancellationResult.Failure("Order already cancelled")
        }

        // 3. Release reserved stock
        val reservations = order.items.map { item ->
            StockReservation(item.productId, item.quantity, reserved = true)
        }
        stockService.releaseStock(reservations)

        // 4. Refund payment if exists
        if (order.transactionId != null) {
            val refundSuccess = paymentService.refundPayment(order.transactionId)
            if (!refundSuccess) {
                return CancellationResult.Failure("Refund failed")
            }
        }

        // 5. Cancel shipment if tracking exists
        if (order.trackingNumber != null) {
            shippingService.cancelShipment(order.trackingNumber)
        }

        // 6. Update order status
        orderRepository.updateStatus(orderId, OrderStatus.CANCELLED)

        return CancellationResult.Success
    }
}

